File: src/api/ffmpeg.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export interface FfmpegStatus {
  available: boolean;
  path: string | null;
  /** どこで見つけたか。 */
  source: "cache" | "bundled" | "path" | "none";
  /** 自動ダウンロードに対応しているか（Windows のみ）。 */
  canDownload: boolean;
}

export type FfmpegProgress =
  | { kind: "start" }
  | { kind: "download"; received: number; total: number }
  | { kind: "extract" }
  | { kind: "done" }
  | { kind: "error"; message: string };

/** 変換用 ffmpeg の現在の状態を取得する。 */
export async function getFfmpegStatus(): Promise<FfmpegStatus> {
  return invoke("get_ffmpeg_status");
}

/** ffmpeg を取得する（既にあればその場所を返す）。進捗は onFfmpegProgress で。 */
export async function downloadFfmpeg(): Promise<string> {
  return invoke("download_ffmpeg");
}

export async function onFfmpegProgress(
  cb: (p: FfmpegProgress) => void,
): Promise<UnlistenFn> {
  return listen<FfmpegProgress>("ffmpeg-progress", (e) => cb(e.payload));
}

File: src-tauri/src/commands/ffmpeg.rs
use tauri::AppHandle;

use crate::ffmpeg;

/// 変換用 ffmpeg の現在の状態（場所・取得元・自動DL可否）を返す。
#[tauri::command]
pub fn get_ffmpeg_status(app: AppHandle) -> ffmpeg::FfmpegStatus {
    ffmpeg::status(&app)
}

/// ffmpeg を取得する（既にあればその場所を返す）。進捗は `ffmpeg-progress` で配信。
#[tauri::command]
pub async fn download_ffmpeg(app: AppHandle) -> Result<String, String> {
    let p = ffmpeg::ensure(&app).await?;
    Ok(p.display().to_string())
}
